[MT-1217] Add launchOption and onReady

sebastian-krajna opened the pull request:
- Updated Branch SDK dependency to version 3.12.1 in Cartfile and related resolved files.
- Changed references from `Branch` to `BranchSDK` across multiple source files for consistency.
- Bumped TealiumBranch version to 2.0.0 in podspec and updated corresponding dependencies.
- Enhanced initialization logic in `BranchInstance` to support launch options.
- Improved test coverage for Branch functionalities, including new onReady handling and event processing.

sebastian-krajna committed:
Update Branch SDK integration and enhance command handling

- Added `onReady` method to `BranchCommand` protocol for better initialization handling.
- Updated `initialize` method to accept `launchOptions` in `BranchInstance` and `BranchRemoteCommand`.
- Modified `TealiumHelper` to configure `BranchRemoteCommand` with launch options.
- Added debug flag support in `branch.json` for improved logging.
- Updated `.gitignore` to include `.vscode/` and `*.xcframework.zip`.
- Bumped Tealium dependency version in `Package.resolved` to 2.17.0.

sebastian-krajna committed:
Update dependencies and versioning for TealiumBranch integration

- Updated Branch SDK dependency to version 3.12 and Tealium Swift to version 2.17 in Cartfile and related resolved files.
- Bumped TealiumBranch version to 2.0.0 in podspec and updated corresponding dependencies.
- Enhanced publish script for better error handling and added support for warnings during pod trunk push.
- Updated README copyright year to 2025 and fixed formatting in various scripts for consistency.

sebastian-krajna committed:
Refactor Branch integration to use BranchSDK

- Updated import statements from `Branch` to `BranchSDK` across multiple source files for consistency.
- Modified project configuration to reflect the change from `Branch.xcframework` to `BranchSDK.xcframework`.
- Removed unused method `getBranchInstance` from `TealiumHelper` to clean up the codebase.
- Added a `.DS_Store` file to the example project directory.

sebastian-krajna committed:
Update Branch SDK integration and enhance functionality

- Updated Branch SDK dependency to version 3.12.1 in Package.swift and Package.resolved.
- Changed references from `Branch` to `BranchSDK` in source files for consistency.
- Added new properties and methods in Branch-related classes to support enhanced event handling and deep linking features.
- Updated .gitignore to include Carthage directory.
- Improved test coverage for Branch functionalities, including event properties and command handling.

sebastian-krajna committed:
Refactor Branch constants and enhance event handling

- Corrected typos in `BranchConstants` for `errorPrefix` and `eventDescription`.
- Added new property `enableDMACompliance` to `BranchConstants`.
- Improved type safety in `TypeChecker` for numeric conversions.
- Updated event handling in `BranchRemoteCommand` to use corrected constants.
- Enhanced `AppDelegate` to handle deep links and push notifications with Tealium.
- Updated example project to reflect new event structures and improved UI layout.
- Enhanced test cases to validate changes in event properties and command handling.

sebastian-krajna committed:
Update Branch configuration with placeholder keys in JSON and Info.plist

- Replaced the placeholder for `branch_dev_key` in `branch.json` with "YOUR_BRANCH_KEY_HERE".
- Added a new section in `TealiumBranchExample-Info.plist` for Branch configuration, including placeholders for test and live branch keys.

sebastian-krajna committed:
Downgrade version and clean up comments in Branch files

- Changed version from "3.0.0" to "2.0.0" in `BranchConstants` and test files.
- Removed commented-out sections in `BranchInstance` and `BranchRemoteCommand` for improved code clarity.
- Enhanced type safety in `TypeChecker` by removing unnecessary comments.

sebastian-krajna committed:
Enhance custom data handling in BranchEvent

- Updated the logic for merging custom data in `BranchEvent` to ensure new values are added without overwriting existing ones. This improves data integrity and allows for more flexible event customization.

sebastian-krajna committed:
Implement push notification handling in Branch SDK

- Added support for handling push notifications in `BranchInstance` and `BranchRemoteCommand`.
- Introduced new constants in `BranchConstants` for push notification user info.
- Updated `AppDelegate` to track push notification events with Tealium.
- Enhanced test coverage for push notification handling in `BranchRemoteCommandTests` and `MockBranchInstance`.

sebastian-krajna committed:
Update Branch SDK integration and clean up project structure

- Changed references from `BranchSDK` to `Branch` across multiple source files for consistency.
- Updated dependency versions in `Cartfile` and `Cartfile.resolved` to reflect the new Branch SDK version 1.40.
- Adjusted `.gitignore` to include `.vscode/` and `Carthage/` directories.
- Refactored `TealiumBranch.podspec` to align with the new versioning and dependencies.
- Enhanced test coverage for Branch functionalities, ensuring compatibility with the updated SDK.

sebastian-krajna committed:
Update .gitignore and remove unnecessary files

- Added `.xcframework.zip`, `.vscode/`, and `Carthage/` to `.gitignore` for better project cleanliness.
- Deleted `.vscode/settings.json` to remove unused configuration.
- Removed `Carthage/.DS_Store` to clean up binary files from the repository.

sebastian-krajna committed:
Update Branch SDK integration and versioning

- Updated Branch SDK dependency to version 3.12.1 in Cartfile and related resolved files.
- Changed references from `Branch` to `BranchSDK` across multiple source files for consistency.
- Bumped TealiumBranch version to 2.0.0 in podspec and updated corresponding dependencies.
- Enhanced initialization logic in `BranchInstance` to support launch options.
- Improved test coverage for Branch functionalities, including new onReady handling and event processing.

denguzov commented on publish.sh:
```diff
@@ -59,7 +60,7 @@ echo "All checks are passed, ready to release to CocoaPods"
 select yn in "Yes" "No"; do
     case $yn in
-        Yes ) echo "Ok, running \"pod trunk push\" now."; pod trunk push; break;;
+        Yes ) echo "Ok, running \"pod trunk push\" now."; pod trunk push --allow-warnings; break;;
```
why are we allowing warnings here?

denguzov commented on publish.sh:
```diff
@@ -10,10 +11,10 @@ else
 
 podspecFile=$(<TealiumBranch.podspec)
-podspecRegex="^.*spec.version[[:space:]]*\= \"([0-9\.]*)\""
+podspecRegex="^.*s.version[[:space:]]*\= \"([0-9\.]*)\""
```
Looking at podspec, is this even gonna work?

sebastian-krajna replied:
oh my bad, I copied it from facebook

denguzov commented on TealiumBranchExample/.DS_Store:
I believe, this file should be git ignored.

denguzov commented on Tests/TealiumBranchTests/BranchInstanceTests.swift:
```diff
@@ -8,14 +8,14 @@
     var branchCommand: BranchRemoteCommand!
 
-    override func setUp() {
+    override func setUp() { 
```
Why was space added?

denguzov commented on Sources/BranchRemoteCommand.swift:
```diff
@@ -42,7 +51,8 @@ public class BranchRemoteCommand: RemoteCommand {
             case BranchConstants.Commands.initialize:
-                branchInstance.initialize(payload: payload)
+                branchInstance.initialize(payload: payload, launchOptions: self.launchOptions)
+            // TODO: rename it to set identity to reflect function name
```
What is the intention for this TODO item?

sebastian-krajna replied:
After a while I know we shouldn't change it, so it doesn't make sense

denguzov commented on Sources/BranchRemoteCommand.swift:
```diff
@@ -16,23 +16,32 @@ import Branch
+                        return
+                    }
+                    weakSelf?.processRemoteCommand(with: payload)
+                })
```
Why make these indentations here?

sebastian-krajna replied:
Okay, some time ago Enrico recommend me that I should use 'format' inside Xcode, so it should be fix now

denguzov replied:
@Enricoza is it smth with our formatting rules? or smth else, what do you think?

denguzov replied:
@sebastian-krajna okay, I think, I know what the issue is. 
First, currently it's better to use `ctrl+a` and `ctrl+i` (or Xcode -> Editor -> Structure -> Re-indent) for the indentation fixes of selected lines because of...
..Second, we still need to come up with `.swift-format` configuration file later, but before we call that command you're using. Perhaps, it wasn't configured previously since that is a new tooling come with XCode 16.

denguzov replied:
@sebastian-krajna at this moment, probably, you could revert all those formatting 'fixes' by adding that configuration file with line length set to 200 (this is the length we use for core SDK library).

denguzov commented on Sources/BranchRemoteCommand.swift:
```diff
@@ -16,23 +16,32 @@ import Branch
     }
     
-    public init(branchInstance: BranchCommand = BranchInstance(), type: RemoteCommandType = .webview) {
+    public init(type: RemoteCommandType = .webview,
```
Why change the order of first two parameters?

sebastian-krajna replied:
I think it's more possible to change the type than the instance

denguzov replied:
I believe, we should strive for kinda 'reading flow' of Swift. So, it seems more reasonable to me saying smth like:
'initialize branch instance of specific type' (first, we say what is initialized, second, we clarify its type)
You could even do smth like `init(branchInstance: BranchCommand, ofType type: RemoteCommandType)`
There is [a guidance](https://www.swift.org/documentation/api-design-guidelines/#naming) for such cases.

denguzov commented on Sources/BranchInstance.swift:
```diff
@@ -40,10 +42,26 @@ public class BranchInstance: BranchCommand, TealiumRegistration {
+        
+        // Enable logging if requested
+        if let logging = logging as? Bool, logging {
+            // TODO: remove deprecated method
```
What is the intention for this TODO item?

sebastian-krajna replied:
As I remember we wanted to create small PR so I added TODO to remove it next PR

denguzov commented on Sources/BranchInstance.swift:
```diff
@@ -40,10 +42,26 @@ public class BranchInstance: BranchCommand, TealiumRegistration {
+    }
+
+    public func onReady(_ onReady: @escaping () -> Void) {
+        TealiumQueues.secureMainThreadExecution {
```
Why has the main thread to be used?

sebastian-krajna replied:
https://github.com/Tealium/tealium-ios-facebook-remote-command/blob/main/Sources/FacebookInstance.swift#L65

I saw that it's used here, so probably in other onReady functions is also needed. So I assume that this pattern ensures thread safety for TealiumReplaySubject operations and UI callbacks.

denguzov replied:
Shouldn't you be securing same thread execution for publishing as well in such case? 🤔

denguzov commented on Sources/BranchInstance.swift:
```diff
@@ -80,23 +98,23 @@ public class BranchInstance: BranchCommand, TealiumRegistration {
+        branchInstance?.setIdentity(id)
     }
     
+    // TODO: remove deprecated method
```
What is the intention for this TODO item?

sebastian-krajna replied:
As I remember we wanted to create small PR so I added TODO to remove it next PR

denguzov commented on Sources/BranchInstance.swift:
```diff
@@ -40,10 +42,26 @@ public class BranchInstance: BranchCommand, TealiumRegistration {
-            }
+        // Get the Branch instance
+        branchInstance = Branch.getInstance()
+        guard let branchInstance else { return }
```
Why are you making `branchInstance` optional?

sebastian-krajna replied:
I changed the approach to match [firebase](https://github.com/Tealium/tealium-ios-firebase-remote-command/blob/main/Sources/FirebaseInstance.swift)

denguzov replied:
Hmm. How are they similar? In Firebase case methods of `Analytics` singleton object are being called 🤔

denguzov commented on Sources/BranchInstance.swift:
```diff
@@ -80,23 +98,23 @@ public class BranchInstance: BranchCommand, TealiumRegistration {
     
     public func setIdentity(id: String) {
-        Branch.getInstance().setIdentity(id)
+        branchInstance?.setIdentity(id)
```
```suggestion
        onReady {
            Branch.getInstance().setIdentity(id)
        }
```

denguzov replied:
Maybe, we could reuse this pattern as we did in other commands?

denguzov commented on Sources/BranchInstance.swift:
```diff
@@ -26,12 +27,13 @@ public protocol BranchCommand {
 
 public class BranchInstance: BranchCommand, TealiumRegistration {
     
+    private var branchInstance: Branch?
```
Why do we need this instance variable now? And why is it optional?

denguzov commented on TealiumBranchExample/TealiumBranchExample/TealiumHelper.swift:
```diff
@@ -5,42 +5,54 @@
+    
+    func configure(with launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
+        // Create Branch remote command with launch options
+        branchRemoteCommand = BranchRemoteCommand(
+            type: .local(file: "branch", bundle: Bundle.main),
+            launchOptions: launchOptions
+        )
         
-        config.addRemoteCommand(branchRemoteCommand)
+        // Set up initialization callback using Facebook pattern
```
Looks like redundant comments. AI leftovers?

denguzov commented on Sources/BranchInstance.swift:
```diff
@@ -40,10 +42,26 @@ public class BranchInstance: BranchCommand, TealiumRegistration {
+            branchInstance.enableLogging()
+        }
+        
+        // Initialize session with launch options
+        branchInstance.initSession(launchOptions: launchOptions ?? [:])
+        
+        // Publish ready state following Facebook pattern
```
Looks like redundant comments. AI leftovers?

denguzov commented on Cartfile:
```diff
@@ -1,2 +1,2 @@
-github "BranchMetrics/ios-branch-deep-linking" ~> 1.40
-github "tealium/tealium-swift" ~> 2.12
+github "BranchMetrics/ios-branch-deep-linking" ~> 3.12
```
Don't forget to update the repo name here 😉

sebastian-krajna committed:
Merge commit '5a01d4e0ef79d1a825641570b45ed23f88ccb43e'

denguzov commented on Sources/BranchExtensions.swift:
Were there any formatting issues with this file?

sebastian-krajna replied:
I just use 'Xcode -> Editor -> Structure -> Format File with ‘swift-format’'

denguzov commented on Sources/BranchInstance.swift:
```diff
@@ -25,51 +29,79 @@ public protocol BranchCommand {
-            branchUniversalObject.addProperties(properties: universalObjectSettings)
+        let universalObjectSettings = parameters[
+            BranchConstants.EventKeys.branchUniversalObjectProperties
+        ]
+        if let universalObjectSettings = universalObjectSettings
+            as? [String: Any]
+        {
+            branchUniversalObject.addProperties(
+                properties: universalObjectSettings
+            )
         }
-        
-        let metadata = parameters[BranchConstants.EventKeys.branchMetadataProperties]
+
+        let metadata = parameters[
+            BranchConstants.EventKeys.branchMetadataProperties
+        ]
```
What's wrong with previous formatting? 🤔

sebastian-krajna replied:
I just use 'Xcode -> Editor -> Structure -> Format File with ‘swift-format’'

denguzov commented on Sources/BranchRemoteCommand.swift:
```diff
@@ -15,105 +16,109 @@ import Branch
+                )
             case BranchConstants.Commands.setUserId:
-                guard let id = payload[BranchConstants.EventKeys.userId] as? String else {
+                guard
```
Why are there so much similar formatting 'fixes' ? 🤔

sebastian-krajna replied:
I just use 'Xcode -> Editor -> Structure -> Format File with ‘swift-format’'

sebastian-krajna committed:
- Wrapped the `_onReady.publish()` call in `BranchInstance` with a thread-safe execution block to ensure consistent behavior during publishing.
- Moved the `type` parameter in `BranchRemoteCommand` constructor to the correct position for improved clarity and consistency.

sebastian-krajna committed:
Remove deprecated `setOptOut` method from `BranchCommand` protocol and `BranchInstance` class. Update logging method to use the current `Branch.enableLogging()` implementation. Improve safety in `TealiumHelper` by conditionally adding `branchRemoteCommand`. Clean up `MockBranchInstance` by removing unused `setOptOut` method and related count variable.

sebastian-krajna commented:
Closing to recreate from scratch.
